File: Go/problem0002.go
package main

import "fmt"

func fibonacci() func() int {
	last1, last2 := 0, 1
	return func() (fib int) {
		fib = last1 + last2
		last1, last2 = fib, last1
		return
	}
}

func main() {
	sum := 0
	f := fibonacci()
	r := f()
	for ; r <= 4000000; r = f() {
		if r%2 == 0 {
			sum += r
		}
	}
	fmt.Println(sum)
}

File: Rust/problem0002.rs
fn fib(n: u32) -> u32 {
    match n {
        1 => 1,
        2 => 2,
        _ => fib(n - 1) + fib(n - 2)
    }
}

fn main() {
    let mut total = 0;
    let mut i = 1;

    loop {
        let res = fib(i);
        if res > 4_000_000 {
            break;
        }
        if res % 2 == 0 {
            total += res;
        }
        i += 1;
    }
    println!("sum: {}", total);
}
